File: src/py/mat3ra/mode/tree.py
from typing import Any, Dict, List, Optional

from mat3ra.esse.models.core.primitive.slugified_entry import SlugifiedEntry
from mat3ra.standata.data.model_tree import MODEL_NAMES, MODEL_TREE
from mat3ra.standata.data.models_tree_config_by_application import (
    models_tree_config_by_application as MODELS_TREE_CONFIG_BY_APPLICATION,
)


def get_pseudopotential_types_from_tree() -> List[str]:
    methods_tree = MODEL_TREE.get("dft", {}).get("gga", {}).get("methods", {})
    return methods_tree.get("pseudopotential", [])


def tree_slug_to_named_object(slug: str) -> SlugifiedEntry:
    name = MODEL_NAMES.get(slug, slug)
    return SlugifiedEntry(slug=slug, name=name)


def get_tree_by_application_name_and_version(name: str, version: str) -> Dict[str, Any]:
    # TODO: add logic to filter by version when necessary
    return MODELS_TREE_CONFIG_BY_APPLICATION.get(name, {})


def get_default_model_type_for_application(application: Dict[str, Any]) -> Optional[str]:
    name = application.get("name")
    if not name:
        return None
    tree = get_tree_by_application_name_and_version(name, application.get("version", ""))
    keys = list(tree.keys())
    return keys[0] if keys else None

File: src/js/tree.ts
import {
    type DFTModelSchema,
    type MLModelSchema,
    type UnknownModelSchema,
    ApplicationSchema,
    SlugifiedEntry,
} from "@mat3ra/esse/dist/js/types";
import MODELS_TREE_CONFIG_BY_APPLICATION from "@mat3ra/standata/dist/js/runtime_data/models/modelsTreeConfigByApplication.json";
import MODEL_TREE_DATA from "@mat3ra/standata/dist/js/runtime_data/models/modelTree.json";
import lodash from "lodash";

import type { ModelTree } from "./types";

export const { MODEL_TREE, MODEL_NAMES } = MODEL_TREE_DATA;

export const METHODS = {
    pseudopotential: "pseudopotential",
    localorbital: "localorbital",
    unknown: "unknown",
} as const;

export const getPseudopotentialTypesFromTree = (): string[] => {
    const dftTree = MODEL_TREE.dft;
    const firstBranch = Object.values(dftTree)[0];
    return firstBranch?.methods?.pseudopotential || [];
};

export const getDFTFunctionalsFromTree = (): string[] => {
    return Object.keys(MODEL_TREE.dft);
};

export const getDFTFunctionalsByApproximation = (
    approximation: "gga" | "lda" | "hybrid" | "other",
): string[] | undefined => {
    const dftTree = MODEL_TREE.dft;
    const branch = dftTree[approximation];
    return branch?.functionals;
};

export const treeSlugToNamedObject = (modelSlug: string): SlugifiedEntry => {
    return {
        slug: modelSlug,
        name: lodash.get(MODEL_NAMES, modelSlug, modelSlug),
    };
};

export const getTreeByApplicationNameAndVersion = ({
    name,
}: Pick<ApplicationSchema, "name" | "version">): ModelTree | undefined => {
    // TODO: add logic to filter by version when necessary
    if (!(name in MODELS_TREE_CONFIG_BY_APPLICATION)) {
        return undefined;
    }
    return MODELS_TREE_CONFIG_BY_APPLICATION[
        name as keyof typeof MODELS_TREE_CONFIG_BY_APPLICATION
    ];
};

export function getDefaultModelTypeSubtypeForApplication(
    application: Pick<ApplicationSchema, "name" | "version">,
) {
    const tree = getTreeByApplicationNameAndVersion(application);

    if (!tree) {
        throw new Error(
            `getDefaultModelTypeForApplication: tree not found for application ${application.name}`,
        );
    }

    const type = Object.keys(tree)[0];
    const subtype = Object.keys(tree[type] || {})[0] || "unknown";

    return { type, subtype } as
        | Pick<DFTModelSchema, "type" | "subtype">
        | Pick<MLModelSchema, "type" | "subtype">
        | Pick<UnknownModelSchema, "type" | "subtype">;
}
